EIPS/eip-2315.md

```diff
@@ -42,69 +64,91 @@ This is the smallest possible change that provides native subroutines without br
+|    0  |      PUSH1  |    3 |        [] |        [] |
+|    2  |    JUMPSUB  |    8 |       [4] |        [] |
+|    4  |   BEGINSUB  |    1 |        [] |      [ 2] |
+|    5  |  RETURNSUB  |    2 |        [] |      [ 2] |
+|    3  |       STOP  |    0 |        [] |        [] |
+
+### Two levels of subroutines
+
+This should execute fine, going into one two depths of subroutines
+Bytecode: `0x6800000000000000000cb300b56011b3b7b5b7`
+
+|  Pc   |      Op     | Cost |   Stack   |   RStack  |
+|-------|-------------|------|-----------|-----------|
+|    0  |      PUSH9  |    3 |        [] |        [] |
+|   10  |    JUMPSUB  |    8 |      [12] |        [] |
+|   12  |   BEGINSUB  |    1 |        [] |      [10] |
+|   13  |      PUSH1  |    3 |        [] |      [10] |
+|   15  |    JUMPSUB  |    8 |      [17] |      [10] |
+|   17  |   BEGINSUB  |    1 |        [] |   [10,15] |
+|   18  |  RETURNSUB  |    2 |        [] |   [10,15] |
+|   16  |  RETURNSUB  |    2 |        [] |      [10] |
+|   11  |       STOP  |    0 |        [] |        [] |
+
+
+### Failure 1: invalid jump
+
+This should fail, since the given `location` is outside of the code-range. The code is the same as previous example, 
+except that the pushed `location` is `0x01000000000000000c` instead of `0x0c`.
+
+Bytecode: `0x6801000000000000000cb300b56011b3b7b5b7`
+
+|  Pc   |      Op     | Cost |   Stack   |   RStack  |
+|-------|-------------|------|-----------|-----------|
+|    0  |      PUSH9  |    3 |        [] |        [] |
+|   10  |    JUMPSUB  |    8 |[18446744073709551628] |        [] |
+
 ```
-                      data  return
-offset step opcode    stack stack
-0      0    PUSH1 3   []    []
-1      1    JUMPSUB   [3]   [1]
-2      4    STOP      []    [1]
-3      2    BEGINSUB  []    [1]
-4      3    RETURNSUB []    []
-```
-The above code should terminate after 4 steps with an empty stack.
-```
-                      data  return
-offset step opcode    stack stack
-0      0    PUSH1 2   []    []
-1      1    JUMPSUB   [2]   [1]
-2      2    BEGINSUB  []    [1]
-3      3    RETURNSUB []    []
+Error: at pc=10, op=JUMPSUB: evm: invalid jump destination
 ```
-The above code should terminate after 4 steps with an empty stack.
+
+### Failure 2: shallow `return_stack`
+
+This should fail at first opcode, due to shallow `return_stack`
+
+Bytecode: `0xb75858` (`RETURNSUB`, `PC`, `PC`)
+
+
+|  Pc   |      Op     | Cost |   Stack   |   RStack  |
+|-------|-------------|------|-----------|-----------|
+|    0  |  RETURNSUB  |    2 |        [] |        [] |
+
 ```
-                      data  return
-offset step opcode    stack stack
-0      0    PUSH1 3   []    []
-1      1    JUMPSUB   [3]   [1]
-2      8    STOP      []    []
-3      2    BEGINSUB  []    [1]
-4      3    PUSH1 7   []    [1]
-5      4    JUMPSUB   [7]   [1,5]
-6      7    RETURNSUB []    [1]
-7      5    BEGINSUB  []    [1]
-8      6    RETURNSUB []    [1]
+Error: at pc=0, op=RETURNSUB: evm: invalid retsub
```

Comment:
I know there is no firm convention around EVM errors yet (even though that would be nice!), but I would probably call this a stack underflow instead of `invalid retsub`.

Reply:
We already have a stack underflow, so I wanted a different message. However, it's an implementation detail in geth

Reply:
okok, if the intention is to make a new error msg, then that seems appropriate.